aten/src/ATen/native/ScatterGather.cpp

```diff
@@ -0,0 +1,82 @@
+#include <tuple>
+#include <algorithm>
+#include <ATen/ATen.h>
+#include <c10/core/WrapDimMinimal.h>
+
+namespace at { namespace native {
+
+Tensor & gather_out(Tensor & result, const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
+  return at::_gather_out(result, self, dim, index, sparse_grad);
+}
+
+Tensor gather(const Tensor & self, int64_t dim, const Tensor & index, bool sparse_grad) {
+  return at::_gather(self, dim, index, sparse_grad);
+}
+
+Tensor & scatter_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  return at::_scatter_(self, dim, index, source);
+}
+
+Tensor & scatter_(Tensor & self, int64_t dim, const Tensor & index, Scalar value) {
+  return at::_scatter_(self, dim, index, value);
+}
+
+Tensor scatter(const Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  Tensor ret = self.clone().contiguous();
+  return at::_scatter_(ret, dim, index, source);
+}
+
+Tensor scatter(const Tensor & self, int64_t dim, const Tensor & index, Scalar value) {
+  Tensor ret = self.clone().contiguous();
+  return at::_scatter_(ret, dim, index, value);
+}
+
+Tensor & scatter_add_(Tensor & self, int64_t dim, const Tensor & index, const Tensor & source) {
+  if (index.numel() == 0) {
```

Comment:
where is this code coming from?  It seems new, why?

Reply:
It comes from https://github.com/zasdfgbnm/pytorch/blob/new-gather-broadcast-123/test/test_torch.py#L8139
```python
# test for empty index, should be a no-op
idx = cast(torch.LongTensor())
actual = getattr(base.clone(), method)(dim, idx, src)
self.assertEqual(actual, base, 0)
```
This is necessary for current implementation of scalar source: if not, then the check at line 41 will be true, then line 44 `source_sizes[dim] = index.size(dim);` will fail.